consensus/src/pos/target.rs

```diff
@@ -0,0 +1,477 @@
+
+    // Determine net upgrade version that current block belongs to.
+    // Across versions config parameters might change so it's invalid to mix versions for target calculations
+    let (_, net_version) = chain_config
+        .net_upgrade()
+        .version_at_height(block_height)
+        .expect("NetUpgrade must've been initialized");
+    let net_version_range = chain_config
+        .net_upgrade()
+        .height_range(net_version)
+        .expect("NetUpgrade must've been initialized");
+
+    // Get timestamps from the history but make sure they belong to the same consensus version
+    let block_times = history_iter
+        .take(pos_config.blocks_count_to_average())
+        .filter(|block_index| net_version_range.contains(&block_index.block_height()))
+        .map(|block_index| block_index.block_timestamp().as_int_seconds())
+        .collect::<Vec<_>>();
+
+    // At least 2 items are required to calculate difference
+    debug_assert!(block_times.len() >= 2);
+
+    // Block times are taken from history so they are sorted backwards
+    let block_diffs = block_times
+        .iter()
+        .tuple_windows::<(&u64, &u64)>()
+        .map(|t| t.0 - t.1)
+        .collect::<Vec<_>>();
+
+    let average = block_diffs.iter().sum::<u64>() / block_diffs.len() as u64;
+    Ok(average)
+}
+
+fn calculate_new_target(
+    pos_config: &PoSChainConfig,
+    prev_target: &Uint256,
+    average_block_time: u64,
+) -> Result<Compact, ConsensusPoSError> {
+    let average_block_time = Uint256::from_u64(average_block_time);
+    let target_block_time = Uint256::from_u64(pos_config.target_block_time().as_secs());
+
+    // TODO: limiting factor (mintlayer/mintlayer-core#787)
+    let new_target = *prev_target / target_block_time * average_block_time;
+
+    if new_target > pos_config.target_limit() {
+        Ok(Compact::from(pos_config.target_limit()))
+    } else {
+        Ok(Compact::from(new_target))
+    }
+}
+
+pub fn calculate_target_required(
+    chain_config: &ChainConfig,
+    pos_status: &PoSStatus,
+    block_header: &BlockHeader,
+    block_index_handle: &impl BlockIndexHandle,
+) -> Result<Compact, ConsensusPoSError> {
+    // check if current block is a net upgrade threshold
+    let pos_config = match pos_status {
+        PoSStatus::Threshold { initial_difficulty } => return Ok(*initial_difficulty),
+        PoSStatus::Ongoing { config } => config,
+    };
+
+    let prev_block_id = match block_header.prev_block_id().classify(chain_config) {
+        GenBlockId::Genesis(_) => return Ok(pos_config.target_limit().into()),
+        GenBlockId::Block(id) => id,
+    };
+    let prev_block_index = block_index_handle
+        .get_block_index(&prev_block_id)?
+        .ok_or_else(|| ConsensusPoSError::PrevBlockIndexNotFound(block_header.get_id()))?;
+
+    // check if prev block is a net upgrade threshold
+    match chain_config.net_upgrade().consensus_status(prev_block_index.block_height()) {
```

Comment:
I don't think this check is necessary. 

At the beginning of the function, the initial difficulty can be retrieved from PoS status. Then, if it's NOT initial, then just take the differences and calculate the average block time, and then use the last target... as usual. No need for special treatments.